histomicstk/MinimumModel.py

```diff
@@ -0,0 +1,795 @@
+
+    return cXs, cYs
+
+
+def ScoreContours(I, cXs, cYs):
+    """Scores boundary contours using gradient information. Implemented from
+    the reference below. Each contour is weighted by the average gradient and
+    number of local gradient maxima along its path.
+
+    Parameters:
+    -----------
+    I : array_like
+        An intensity image used for analyzing local minima/maxima and
+        gradients. Dimensions M x N.
+    cXs : list
+        A list of 1D numpy arrays defining the horizontal coordinates of object
+        boundaries.
+    cYs : list
+        A list of 1D numpy arrays defining the vertical coordinates of object
+        boundaries.
+
+    Notes:
+    ------
+    Implemented from the reference below.
+
+    Returns:
+    --------
+    Scores : array_like
+        A 1D array of horizontal coordinates of contour seed pixels for
+        tracing.
+
+    See Also:
+    ---------
+    TraceContours, LabelContour, MinimumModel
+
+    References:
+    -----------
+    .. [1] S. Weinert et al "Detection and Segmentation of Cell Nuclei in
+    Virtual Microscopy Images: A Minimum-Model Approach" in Nature Scientific
+    Reports,vol.2,no.503, doi:10.1038/srep00503, 2012.
+    """
+
+    # initialize output
+    Scores = np.zeros((len(cXs)))
+
+    # generate Sobel filter response from input intensity image 'I'
+    Gradients = ft.sobel(I, mode='mirror')
+
+    # generate local max in 3 x 3 window of Gradients
+    Maxima = ft.maximum_filter(Gradients, size=3, mode='mirror')
+
+    # generate score for each contour
+    for i in np.arange(len(cXs)):
+
+        # get gradient pixels, local max gradient pixels
+        cG = Gradients[cYs[i], cXs[i]]
+        cMax = Maxima[cYs[i], cXs[i]]
+
+        # calculate mean gradient
+        MG = np.sum(np.abs(Gradients[cYs[i], cXs[i]])) / len(cXs[i])
+
+        # calculate gradient fit
+        GF = np.sum(cG == cMax) / len(cXs[i])
+
+        # compute score as product of mean gradient and gradient fit
+        Scores[i] = MG * GF
+
+    return Scores
+
+
+def LabelContour(Shape, cXs, cYs, Scores):
+    """Constructs a label image from scored contours. Masks for contours with
+    low priority/score are placed first into the label image and then are
+    overwritten by higher priority contours.
+
+    Parameters:
+    -----------
+    Shape : tuple
+        The shape tuple of the desired label image (height, width).
+    cXs : list
+        A list of 1D numpy arrays defining the horizontal coordinates of object
+        boundaries.
+    cYs : list
+        A list of 1D numpy arrays defining the vertical coordinates of object
+        boundaries.
+    Scores : array_like
+        A 1D array of horizontal coordinates of contour seed pixels for
+        tracing.
+
+    Notes:
+    ------
+    Can produce a large number of thin "halo" objects surrouding the objects
+    with higher scores. These can be removed by filtering object width in the
+    resulting label image.
+
```

Comment:
When a portion of an object with lower score is overwritten by an object with higher score, maybe it would be better to remove/erase the object with lower score completely? What would be the use case that prevents us from doing this?